### DISCUSSION TITLE
Metals and the Google Summer of Code (GSOC) 2021

### DISCUSSION THREAD

**Discussion by ckipp01** [scalameta/metals#2593] [General]
The [Scala Center](https://scala.epfl.ch/) is leading this year's Google Summer of Code, and we're happy to say that our Metals will be a part of it! If you are interested in joining, please check out more info here: [https://summerofcode.withgoogle.com/organizations/5449337515016192/](https://summerofcode.withgoogle.com/organizations/5449337515016192/).

If you have any questions, don't hesitate to ask.

**Comment by tanishiking** [comment]
Hi @ckipp01 and @tgodzik, thank you so much for participating in GSoC and making great project ideas!
I'm Rikito Taniguchi a.k.a. @tanishiking, and I would love to apply to GSoC and working on scalameta and metals intensively in the period.

Looking at [ideas list](https://github.com/scalacenter/GoogleSummerOfCode2021/blob/main/README.md), I'm interested in [Add synthetics and symbol information for semanticdb in Scala 3](https://github.com/scalacenter/GoogleSummerOfCode2021/blob/main/README.md#Add-synthetics-and-symbol-information-for-semanticdb-in-Scala-3) project in particular,
because my current primary interest lies in improving developers' experience for scala3, and I believe more coverage in semanticdb for scala3 leads to better tooling support of metals, scalafix, and other semanticdb-based developer tools.

Is it possible to elaborate on this project overview? 
I'm willing to delve into what kind of information missing in semanticdb extracted from scala3 and what problems are happening because of this, but it'd be awesome if there's some guidance :)
(I guess I'm familiar with [semanticdb specification](https://scalameta.org/docs/semanticdb/specification.html) and I know [here](https://github.com/dotty-staging/dotty/blob/bd906c40b735186a7e28853c896d6284d5f54d72/compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala#L30) is the endpoint of semanticdb extracting phase in scala3 compiler).

(In addition to semanticdb issues, `"Add missing LSP extensions to Metals"` and `"Automatic imports of implicits"` looks also interesting!)

Btw, if there's another better place to contact (like emailing or discord) let me know, I'm happy to move :)

**Reply by tgodzik** [reply]
As you said semanticdb is used in a lot of places and we also use the information about the signatures of symbols defined in particular source files. Let's take as an example a simple file like:
```
class Simple {
  val a = 123
  a + a
}
```

symbols section will look like:

<details><summary>Scala 2</summary>
<p>

```scala
 Vector(
        SymbolInformation(
          "_empty_/Simple#",
          SCALA,
          CLASS,
          0,
          "Simple",
          ClassSignature(
            Some(Scope(Vector(), Vector())),
            Vector(TypeRef(Empty, "scala/AnyRef#", Vector())),
            Empty,
            Some(Scope(Vector("_empty_/Simple#`<init>`().", "_empty_/Simple#a."), Vector()))
          ),
          Vector(),
          PublicAccess()
        ),
        SymbolInformation(
          "_empty_/Simple#`<init>`().",
          SCALA,
          CONSTRUCTOR,
          8192,
          "<init>",
          MethodSignature(Some(Scope(Vector(), Vector())), Vector(Scope(Vector(), Vector())), Empty),
          Vector(),
          PublicAccess()
        ),
        SymbolInformation(
          "_empty_/Simple#a.",
          SCALA,
          METHOD,
          1024,
          "a",
          MethodSignature(
            Some(Scope(Vector(), Vector())),
            Vector(),
            TypeRef(Empty, "scala/Int#", Vector())
          ),
          Vector(),
          PublicAccess()
        )
      ),
```

compared to Scala 3 which is missing all information about signatures. 
</p>
</details>

<details><summary>Scala 3</summary>
<p>

```scala
      Vector(
        SymbolInformation("_empty_/Simple#", SCALA, CLASS, 0, "Simple", Empty, Vector(), Empty),
        SymbolInformation(
          "_empty_/Simple#`<init>`().",
          SCALA,
          CONSTRUCTOR,
          8192,
          "<init>",
          Empty,
          Vector(),
          Empty
        ),
        SymbolInformation("_empty_/Simple#a.", SCALA, METHOD, 1024, "a", Empty, Vector(), Empty)
      ),
```

</p>
</details>

Without the signatures Metals is unable to support `implementations` and `type decorations` (maybe something more.)

The first step would be to include the signatures when generating semanticdb in the scala 3 compiler. Another step would be adding information about synthetic symbols being used in the code (added types for generic methods, apply methods used without `apply` etc.). Some examples can be found in here: https://github.com/scalameta/metals/blob/main/metals/src/main/scala/scala/meta/internal/decorations/SemanticdbTreePrinter.scala#L160

I can explain it further if needed :)

**Reply by tanishiking** [reply]
@tgodzik Thank you for the information!

I took a glance at scalameta and metals, I guess I understand what we need to do and what we can do with that information. It looks so interesting, and I'd like to apply to this project!

As the next step, I'm going to investigate more what exactly information missing in scala3 in comparison to scala2, by comparing semanticdb generated from scala2 and scala3. And start to write a proposal draft.

Let me ask some more questions: 🙏 

- How did you print those `SymbolInformation`s you pasted?
  - Maybe, by printing symbols information?
    - [here in scala2](https://github.com/scalameta/scalameta/blob/2683e4be3c7ca7478ab2243f46b3d622df7d6079/semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala#L765)
    - [around here in scala3](https://github.com/tanishiking/dotty/blob/64aa7cfb0a0a245261c490e268052880a07f5c00/compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala#L594)
- While data structures for semanticdb is generated from [protocol buffer schema](https://github.com/scalameta/scalameta/blob/2683e4be3c7ca7478ab2243f46b3d622df7d6079/semanticdb/semanticdb/semanticdb.proto#L362-L365) in scala2, the data structure in scala3 looks different (though they don't look like purely hand-crafted).
  - How did we developed data structure ( like [SymbolInformation](https://github.com/lampepfl/dotty/blob/64aa7cfb0a0a245261c490e268052880a07f5c00/compiler/src/dotty/tools/dotc/semanticdb/SymbolInformation.scala#L128)) in scala3?
    - Did we generated the code from somewhere? or partially copy and pasted from scala2's? or purely hand-crafted?
  - Maybe we should ask @bishabosha ?

<details>
<summary>Some notes about Signatures and Synthetics</summary>

## What features are missing in scala3 because of MethodSignature and Synthetics
### Signatures
#### who use this
- At a glance, in metals, `(Method|Type|Class|Value)Signature` information is consumed by `Implementation` and `decoration` as you mentioned.
- Therefore, `textDocument/implementation` and `show inferred type / show implicit` features are infeasible in scala3.
- Besides, goto supermethods feature and `textDocuemnt/rename` maybe affected, since they depend on `implementionProvider`.

#### where in scala2 generate this info
[SymbolInformationOps](https://github.com/scalameta/scalameta/blob/2683e4be3c7ca7478ab2243f46b3d622df7d6079/semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/SymbolInformationOps.scala#L251) is the endpoint to generate `Signature` section in SymbolInformation.

---

### Synthetics
https://scalameta.org/docs/semanticdb/specification.html#synthetic

#### who use this
- In metals, this information is used by `show inferred type / show implicit` features, (while scalafix's built-in rules don't look like consuming this info).

#### where in scala2 generate this info
- In semanticdb in scala2, Synthetics information is collected here [TextDocumentOps](https://github.com/scalameta/scalameta/blob/2683e4be3c7ca7478ab2243f46b3d622df7d6079/semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala#L59), and [SyntheticPrinter](https://github.com/scalameta/scalameta/blob/2683e4be3c7ca7478ab2243f46b3d622df7d6079/semanticdb/metap/src/main/scala/scala/meta/internal/metap/SyntheticPrinter.scala#L6), prints that information into semanticdb.
- However, `Synthetic` information is completely missing in semanticdb extracted from scala3 now.
  - actually there's no `synthetics` field in scala3 https://github.com/tanishiking/dotty/blob/64aa7cfb0a0a245261c490e268052880a07f5c00/compiler/src/dotty/tools/dotc/semanticdb/TextDocument.scala#L10-L17

</details>

**Reply by tgodzik** [reply]
> How did you print those SymbolInformations you pasted?

I uses pprint, just pack anything into `pprint.log(...)` and it will pretty print it.

> How did we developed data structure ( like SymbolInformation) in scala3?

I think that was partly copied from scalameta, we might also need to update it. @bishabosha might know more in that regard for sure.

>Some notes about Signatures and Synthetics

Looks correct. From the look of it any my personal experience it looks like Signatures are much more important. We should start with that - synthetics are really nice to have, but if we run out of time we might skip it.

**Reply by bishabosha** [reply]
Hey, so to generate all the code in the `dotty.tools.dotc.semanticdb` package that derives from the protobuf scheme, we just copied the code generated by ScalaPB for scalameta and trimmed it down and also annotate variables with `@scala.annotation.internal.sharable`, you also need to replace occurrences of protobuf stream types with `SemanticdbInputStream` or `SemanticdbOutputStream` - this is because we copied the implementation of protobuf streams to avoid a dependency, but renamed it

**Reply by bishabosha** [reply]
to debug the semanticdb as you write it, you can take a look at https://github.com/lampepfl/dotty/blob/bf98ec588f83e2d1b83f30b1e4be802e66939463/compiler/src/dotty/tools/dotc/semanticdb/Tools.scala#L49

**Reply by tanishiking** [reply]
@tgodzik Thanks for checking my notes. `pprint` seems useful :)

>  From the look of it any my personal experience it looks like Signatures are much more important. We should start with that 

Agree, sounds good 👍 

---

@bishabosha Thank you for the information!
Though It might be ideal to embed the code-generation to compilation workflow in the future, copy-and-pasting the code seems reasonable 👍 
And thank you for sharing Tools.scala, It's good to hear there's metac in scala3.

**Reply by tanishiking** [reply]
@tgodzik 
Hi, I wrote up a draft proposal for the project, could you review it?
https://docs.google.com/document/d/1YAQy0J9hFUOjUL3OD_6rEuCM6nUYRExQHDm3I2q82cc/edit?usp=sharing

- Planned a schedule based on my rough estimation.
  - I want you to check the order of the tasks.
  - Regarding the time duration for each task, it's just based on my rough estimation. If you feel it's quite different from your estimation, please tell me.
- Mentioned a task to extend SemanticDB's type system so we can express new Scala3 types such as [TypeLambdas](https://dotty.epfl.ch/docs/reference/new-types/type-lambdas.html) and [MatchType](https://dotty.epfl.ch/docs/reference/new-types/match-types.html), what do you think about this?
  - I guess this is needed because the current SemanticDB type system is designed after Scala2's type, and some new Scala3 Types cannot be expressed by the current SemanticDB schema.

**Reply by tgodzik** [reply]
Thanks! I will look at it later today.

**Reply by tgodzik** [reply]
I don't think I have anything to add. It does look pretty good! I think synthetics is good to keep as a stretch goal.

**Comment by twinlensreflex** [comment]
Hello!

I'm interested in submitting a GSOC proposal for [Add missing LSP extensions to Metals](https://github.com/scalacenter/GoogleSummerOfCode2021#add-missing-lsp-extensions-to-metals
).

I've been looking at [Google's student proposal guide](https://google.github.io/gsocguides/student/writing-a-proposal#elements-of-a-quality-proposal) and was wondering if there are any specific rules or templates you have for proposals? The guide says to check.

**Reply by ckipp01** [reply]
Hey @eleanombre!

> I'm interested in submitting a GSOC proposal for Add missing LSP extensions to Metals.

That's great! I think tackling this should be pretty fun, and totally doable. They both will give a really good picture of how Metals handles LSP requests, interacts with the Scala presentation compiler, and also uses [SemanticDB](https://scalameta.org/docs/semanticdb/guide.html). As you are looking possibly looking into either of these, don't hesitate to ask for any pointers about where to look or how things are layed out.

> I've been looking at Google's student proposal guide and was wondering if there are any specific rules or templates you have for proposals? The guide says to check.

So we don't have any specific template or anything to follow. I'd recommend to just follow the process that is outlined. Again as you are preparing your proposal and looking through things, don't hesitate to ask any questions.
